### ISSUE TITLE
Read-locking shouldn't hang if thread has already a write-lock?

### ISSUE THREAD

**Issue by KrzysztofMadejski** [golang/go#30657]
### What version of Go are you using (`go version`)?

go version go1.12 linux/amd64

### Does this issue reproduce with the latest release?

This is the latest release.

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPROXY=""
GORACE=""
GOROOT="/snap/go/3318"
GOTMPDIR=""
GOTOOLDIR="/snap/go/3318/pkg/tool/linux_amd64"
GCCGO="gccgo"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build217834812=/tmp/go-build -gno-record-gcc-switches"

</pre></details>

### What did you do?

Run
```
package main

import (
	"fmt"
	"sync"
)

func main() {
	mutex := &sync.RWMutex{}

	// locking
	fmt.Print("1")
	mutex.Lock()
	fmt.Print("2")
	mutex.RLock()

	// unlocking
	mutex.RUnlock()
	fmt.Print("3")
	mutex.Unlock()
	fmt.Print("4")
}

```

### What did you expect to see?

I expected the program to run fully and output `1234`. I would expect that the `RLock` should not hang because the current thread already have a write lock on that mutex. 

### What did you see instead?

```
12fatal error: all goroutines are asleep - deadlock!

goroutine 1 [semacquire]:
sync.runtime_SemacquireMutex(0xc00001656c, 0xc00000e000)
	/snap/go/3318/src/runtime/sema.go:71 +0x3d
sync.(*RWMutex).RLock(...)
	/snap/go/3318/src/sync/rwmutex.go:50
main.main()
	/home/kmadejski/Projects/go-whatsapp/test.go:15 +0x211

Process finished with exit code 2
```

or 

```
12
```

and the project hanged if run with `-race` flag.

**Comment by kardianos** [comment]
That's not how locks in Go work. Locks in Go know nothing about goroutines or threads, just code calling order. Not can you upgrade or downgrade a RW lock.

This is better.

**Comment by KrzysztofMadejski** [comment]
Yes, I already realized that Go community have a lot of assumptions what is better. Could you point me to rationale explaining why this is better?

**Comment by bcmills** [comment]
> Could you point me to rationale explaining why this is better?

See discussion in #9201, which seems like a prerequisite to even considering this change.

**Comment by bcmills** [comment]
See previously #23513 and #15418.

**Comment by KrzysztofMadejski** [comment]
Thanks!

**Comment by KrzysztofMadejski** [comment]
https://github.com/golang/go/issues/15418#issuecomment-216205544 @cznic 
> What does not work is when the same goroutine calls RLock on a RWMutex when that goroutine already did that without calling RUnlock first.

I skimmed through all the 3 discussion and nowhere I found any rationale supporting that implementation.

My statement against such implementation is that in a complex code (any that have multiple threads in my opinion can be called complex) it is hard to ensure that double locking or upgrading locking won't happen in one thread. Why the programmer should care about it if it can be make easier?

@kardianos 
>  Locks in Go know nothing about goroutines or threads, just code calling order.

So basically you are saying that the most crucial mechanism to synchronize parallelly running threads does not know anything about threads?

**Comment by mvdan** [comment]
> So basically you are saying that the most crucial mechanism to synchronize parallelly running threads does not know anything about threads?

In Go, locks are not crucial, nor part of the language. Those would be channels. Anything that's not part of the language can only use language features, and inspecting which thread/goroutine is running is not part of the language spec for good reason.

**Comment by KrzysztofMadejski** [comment]
That's a good insight. Thank you @mvdan ! I will evaluate how I can use channels instead of mutexes to synchronize then.

**Comment by DisposaBoy** [comment]
@KrzysztofMadejski

Out of curiosity... do you have a more substantial example of a problem that this feature would solve? 

Given your example (and all examples I've ever seen) I don't understand why one would write code like that and it always looked (to me) like:

* code that's hard to reason about for no obvious reason
* obviously a bug
* symptom of poor design/architecture
* no clear ownership
* all the above

**Comment by KrzysztofMadejski** [comment]
Currently I'm extending an open-source tool feature, so it's partly driven by the architectural decisions being taken by the owners. These are not well documented and I feel like there is some technological debt. I'm asking for better documentation and trying to understand the current architecture. 

My case
It is a library to process asynchronously incoming messages to & from a chat provider:
- there is a routine for incoming messages that read & hang from a websocket
- there is a routine for outgoing messages that write to websocket
- there is a routine for keep-alive

Currently I'm implementing close. While closing no messages should be processed. Also before closing the connection all running go routines should exit.  Right now there is a mutex guarding operations on connection and a `connIsOpen` flag.  While closing I'm getting a WLock so no messages are processed (writing/reading from the websocket is guarded by W / RMutex). I guess in this case it was a bug I got a double lock.

I'm more worried about the reconnection process, because it can be triggered externally or from within the message handling functions. 

It's also hard to debug such case of double locking.

**Comment by ianlancetaylor** [comment]
We aren't going to change the way that `sync.RWMutex` behaves, so I'm going to close this issue.

The idea that `RLock` can succeed because the current goroutine already holds `Lock` would require a close tie between mutexes and goroutines that does not exist today.  Perhaps a different type could implement that functionality, but there is no plan to implement that for `sync.RWMutex`.  Or perhaps a different type could implement an upgradeable or downgradeable lock, for a different approach to the problem.

The current `sync.Mutex` and `sync.RWMutex` types are simple and will remain that way.
